archive/walk.go

```diff
@@ -2,18 +2,232 @@ package archive
 
 import (
 	"context"
+	"fmt"
 	"io"
+	"path"
+	"regexp"
+	"sort"
+	"strconv"
 	"strings"
+	"time"
 
 	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/pkg/byteconv"
 	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/pkg/nano"
 	"github.com/brimsec/zq/scanner"
 	"github.com/brimsec/zq/zbuf"
 	"github.com/brimsec/zq/zio"
 	"github.com/brimsec/zq/zio/detector"
 	"github.com/brimsec/zq/zng/resolver"
+	"github.com/segmentio/ksuid"
 )
 
+const (
+	dataDirname = "zd"
+	zarExt      = ".zar"
+)
+
+// A fileKind is the first part of a file name, used to differentiate files
+// when they are listed from the archive's backing store.
+type fileKind string
+
+const (
+	fileKindUnknown fileKind = ""
+	fileKindData             = "d"
+	fileKindSeek             = "ts"
+)
+
+// A dataFile holds archive record data. Only one kind of data file
+// currently exists, representing a file created during ingest.
+type dataFile struct {
+	id   ksuid.KSUID
+	kind fileKind
+}
+
+func newDataFile() dataFile {
+	return dataFile{ksuid.New(), fileKindData}
+}
+
+func (f dataFile) name() string {
+	return fmt.Sprintf("%s-%s.zng", f.kind, f.id)
+}
+
+var dataFileNameRegex = regexp.MustCompile(`d-([0-9A-Za-z]{27}).zng$`)
+
+func dataFileNameMatch(s string) (f dataFile, ok bool) {
+	match := dataFileNameRegex.FindSubmatch([]byte(s))
+	if match == nil {
+		return
+	}
+	id, err := ksuid.Parse(byteconv.UnsafeString(match[1]))
+	if err != nil {
+		return
+	}
+	return dataFile{id, fileKindData}, true
+}
+
+// A seekIndexFile is a microindex whose keys are record timestamps, and whose
+// values are seek offsets into the data file with the same uuid. The name
+// of a seekIndexFile encodes the number of records, and the first & last
+// record timestamps of the corresponding data file. The order of the
+// first & last records matches the archive's data order.
+type seekIndexFile struct {
+	id          ksuid.KSUID
+	first       nano.Ts
+	last        nano.Ts
+	recordCount int64
+}
+
+func (f seekIndexFile) name() string {
+	return fmt.Sprintf("%s-%s-%d-%d-%d.zng", fileKindSeek, f.id, f.recordCount, f.first, f.last)
+}
+
+func (f seekIndexFile) span() nano.Span {
+	return nano.Span{Ts: f.first, Dur: 1}.Union(nano.Span{Ts: f.last, Dur: 1})
+}
+
+var seekIndexNameRegex = regexp.MustCompile(`ts-([0-9A-Za-z]{27})-([0-9]+)-([0-9]+)-([0-9]+).zng$`)
+
+func seekIndexNameMatch(s string) (f seekIndexFile, ok bool) {
+	match := seekIndexNameRegex.FindSubmatch([]byte(s))
+	if match == nil {
+		return
+	}
+	id, err := ksuid.Parse(byteconv.UnsafeString(match[1]))
+	if err != nil {
+		return
+	}
+	recordCount, err := strconv.ParseInt(byteconv.UnsafeString(match[2]), 10, 64)
+	if err != nil {
+		return
+	}
+	firstTs, err := strconv.ParseInt(byteconv.UnsafeString(match[3]), 10, 64)
+	if err != nil {
+		return
+	}
+	lastTs, err := strconv.ParseInt(byteconv.UnsafeString(match[4]), 10, 64)
```

Comment:
Nit: As above, no need for []byte(s) or byteconv.UnsafeString if you use seekIndexNameRegex.FindStringSubmatch.

Reply:
I didn't want to use the regex string functions since we don't need to create the strings; I did change this to use the `byteconv.ParseInt64` function, so it's clearer.

Reply:
Both FindSubmatch and FindStringSubmatch return a slice of slices of their argument, so their allocation behavior is the same.

Reply:
ah, thanks!